### ISSUE TITLE
[JENKINS-25040] Unnecessary SCM actions in multi-configuration job on master

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/matrix-project-plugin#587]
<p>I have a multi-configuration Job that has to run on multiple nodes at the same time. I want the job to checkout from CVS, then do build steps on the CVS content. </p>

<p><b>Current situation</b><br/>
-Multiple slave nodes are assigned to the job. <br/>
-Master is NOt assigned to the job.<br/>
-CVS is accessed by pserver CVSROOT and defined in the job configuration</p>

<p><b>Current behaviour</b><br/>
-Master checks out from CVS, then delegates job to the slave nodes, which also do a CVS checkout, then do the job</p>

<p><b>Expected behaviour</b><br/>
-Master immediately delegates to the assigned nodes without doing a CVS chekout first. Since the master is not an assigned node of the job, it is not supposed to do anything concerning the job.</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=isidius">isidius</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-25040" target="_blank">Unnecessary SCM actions in multi-configuration job on master</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/rsandell.png" title="rsandell's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rsandell">rsandell</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Minor
<li><b>component(s)</b>: matrix-project-plugin
<li><b>resolution</b>: Duplicate
<li><b>resolved</b>: 2014-10-08T20:08:16+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 3
<li><b>imported</b>: 2025-12-06
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>I have a multi-configuration Job that has to run on multiple nodes at the same time. I want the job to checkout from CVS, then do build steps on the CVS content. </p>

<p><b>Current situation</b>
-Multiple slave nodes are assigned to the job. 
-Master is NOt assigned to the job.
-CVS is accessed by pserver CVSROOT and defined in the job configuration</p>

<p><b>Current behaviour</b>
-Master checks out from CVS, then delegates job to the slave nodes, which also do a CVS checkout, then do the job</p>

<p><b>Expected behaviour</b>
-Master immediately delegates to the assigned nodes without doing a CVS chekout first. Since the master is not an assigned node of the job, it is not supposed to do anything concerning the job.</p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>Jenkins 1.583, Windows OS</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/oleg_nenashev.png" title="oleg_nenashev's avatar" /> oleg_nenashev:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/25040?focusedId=213294&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213294">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-25040?focusedId=213294&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213294">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Your request seems to be similar to <a href="https://issue-redirect.jenkins.io/issue/4960" title="Provide a SCMCheckoutStrategy to checkout only once for all matrix entries" class="issue-link" data-issue-key="JENKINS-4960">JENKINS-4960</a>.
Would it resolve your case?</p>
</pre>
</details>

</ul>
</details>

<p>Your request seems to be similar to <a href="https://issue-redirect.jenkins.io/issue/4960" title="Provide a SCMCheckoutStrategy to checkout only once for all matrix entries" class="issue-link" data-issue-key="JENKINS-4960">JENKINS-4960</a>.<br/>
Would it resolve your case?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/25040?focusedId=213316&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213316">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-25040?focusedId=213316&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213316">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>In this case <a href="https://issue-redirect.jenkins.io/issue/14618" title="ignore the checkout on the master" class="issue-link" data-issue-key="JENKINS-14618">JENKINS-14618</a> is a better match.</p>

<p>Is this about saving time/space, or does master have connectivity issues or something else preventing successful checkout?</p>
</pre>
</details>

</ul>
</details>

<p>In this case <a href="https://issue-redirect.jenkins.io/issue/14618" title="ignore the checkout on the master" class="issue-link" data-issue-key="JENKINS-14618">JENKINS-14618</a> is a better match.</p>

<p>Is this about saving time/space, or does master have connectivity issues or something else preventing successful checkout?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>isidius:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/25040?focusedId=213335&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213335">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-25040?focusedId=213335&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213335">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Yes <a href="https://issue-redirect.jenkins.io/issue/14618" title="ignore the checkout on the master" class="issue-link" data-issue-key="JENKINS-14618">JENKINS-14618</a> would fit the description. The nodes of the matrix run on different machines, where the matrix job is to be executed. The master's only job is to distribute the job to the slaves. As Daniel Beck says, the issue <a href="https://issue-redirect.jenkins.io/issue/4960" title="Provide a SCMCheckoutStrategy to checkout only once for all matrix entries" class="issue-link" data-issue-key="JENKINS-4960">JENKINS-4960</a> describes the exact opposite where the master is supposed to do the checkout. </p>

<p>If I follow the idea of <a href="https://issue-redirect.jenkins.io/issue/4960" title="Provide a SCMCheckoutStrategy to checkout only once for all matrix entries" class="issue-link" data-issue-key="JENKINS-4960">JENKINS-4960</a>, I would indeed have a problem since the master is on a slow network subnet, where-as the slaves have direct access to the SCM. Having the master do the checkout and then somehow transfer the data to the slaves would mean a big increase in performance/time loss.</p>

<p>So yes, as Daniel Beck already stated, <a href="https://issue-redirect.jenkins.io/issue/14618" title="ignore the checkout on the master" class="issue-link" data-issue-key="JENKINS-14618">JENKINS-14618</a> describes my point, altho somewhat more compact... <a href="https://issue-redirect.jenkins.io/issue/4960" title="Provide a SCMCheckoutStrategy to checkout only once for all matrix entries" class="issue-link" data-issue-key="JENKINS-4960">JENKINS-4960</a> is not related to my problem.</p>

<p>I intent to save time by directly passing the job to the slaves configured in the job. The master has no actual work to do except hand the job to the slaves and receive the result of the job.</p>
</pre>
</details>

</ul>
</details>

<p>Yes <a href="https://issue-redirect.jenkins.io/issue/14618" title="ignore the checkout on the master" class="issue-link" data-issue-key="JENKINS-14618">JENKINS-14618</a> would fit the description. The nodes of the matrix run on different machines, where the matrix job is to be executed. The master's only job is to distribute the job to the slaves. As Daniel Beck says, the issue <a href="https://issue-redirect.jenkins.io/issue/4960" title="Provide a SCMCheckoutStrategy to checkout only once for all matrix entries" class="issue-link" data-issue-key="JENKINS-4960">JENKINS-4960</a> describes the exact opposite where the master is supposed to do the checkout. </p>

<p>If I follow the idea of <a href="https://issue-redirect.jenkins.io/issue/4960" title="Provide a SCMCheckoutStrategy to checkout only once for all matrix entries" class="issue-link" data-issue-key="JENKINS-4960">JENKINS-4960</a>, I would indeed have a problem since the master is on a slow network subnet, where-as the slaves have direct access to the SCM. Having the master do the checkout and then somehow transfer the data to the slaves would mean a big increase in performance/time loss.</p>

<p>So yes, as Daniel Beck already stated, <a href="https://issue-redirect.jenkins.io/issue/14618" title="ignore the checkout on the master" class="issue-link" data-issue-key="JENKINS-14618">JENKINS-14618</a> describes my point, altho somewhat more compact... <a href="https://issue-redirect.jenkins.io/issue/4960" title="Provide a SCMCheckoutStrategy to checkout only once for all matrix entries" class="issue-link" data-issue-key="JENKINS-4960">JENKINS-4960</a> is not related to my problem.</p>

<p>I intent to save time by directly passing the job to the slaves configured in the job. The master has no actual work to do except hand the job to the slaves and receive the result of the job.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>isidius:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/25040?focusedId=213337&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213337">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-25040?focusedId=213337&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213337">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Linked another issue describing the problem.</p>
</pre>
</details>

</ul>
</details>

<p>Linked another issue describing the problem.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/25040?focusedId=213339&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213339">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-25040?focusedId=213339&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213339">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Duplicates <a href="https://issue-redirect.jenkins.io/issue/14618" title="ignore the checkout on the master" class="issue-link" data-issue-key="JENKINS-14618">JENKINS-14618</a>. Resolving this one as the other is older and has more watchers. Feel free to copy the description why you need this over into a 'Me, too' comment.</p>
</pre>
</details>

</ul>
</details>

<p>Duplicates <a href="https://issue-redirect.jenkins.io/issue/14618" title="ignore the checkout on the master" class="issue-link" data-issue-key="JENKINS-14618">JENKINS-14618</a>. Resolving this one as the other is older and has more watchers. Feel free to copy the description why you need this over into a 'Me, too' comment.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>isidius:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/25040?focusedId=213363&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213363">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-25040?focusedId=213363&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213363">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Issue is duplicate. Necessary information has been copied to comment at primary issue.</p>
</pre>
</details>

</ul>
</details>

<p>Issue is duplicate. Necessary information has been copied to comment at primary issue.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `duplicates` from Jira: <a href="https://github.com/jenkinsci/matrix-project-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-14618%22">JENKINS-14618</a>]</i>
